### ISSUE TITLE
Improve performance of Godot Physics Server 2D/3D active callback events

### ISSUE THREAD

**Issue by Ughuuu** [godotengine/godot-proposals#10389]
### Describe the project you are working on

[Godot Rapier](https://github.com/appsinacup/godot-rapier-physics) Physics Addon.

### Describe the problem or limitation you are having in your project

The 2 functions that take the most time are:
- `step`
- `flush_queries`

Notably the `flush_queries` one takes about as much as the step. This is not right. The flow of the calls for `flush_queries` are as follows:

- The RigidBody2D and RigidBody3D (and other nodes) call the PhysicsServer with `body_set_state_sync_callback`. This sets a callback to be called when this body is active. Then Godot calls back into the body to get some information, such as velocities, positions, sleep state. This happens at Godot side in the `_sync_body_state` function. This function is internal

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I would propose the flush_queries would not flush the queries, but instead a different approach is taken.

Right now flush_queries calls the callbacks, and then it's the nodes job to get what info they need, decoupling things a lot. This is great for architecture design. However for performance it's not as efficient as it could be.

I propose we do this update in a larger batch, rather than one by one.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I propose we have a function `PhysicsServer::space_get_active_bodies(space_rid) -> Array[RID]`. This function returns the RID of all active objects.

Then, we have another function, `PhysicsServer::space_get_bodies_transforms(space_rid, array_bodies_rid) -> Array[Transform]`. This function returns an array of transforms for the bodies.

Then, we have another function, `PhysicsServer::space_get_bodies_linear_velocities(space_rid, array_bodies_rid) -> PackedVectorArray`. This function returns an array of linear velocities for the bodies.

Same for angular velocity and is_sleeping.

The idea here would be we:
- get active body RID's
- iterate through all RID's and match then with actual object. Check which of these need the transform(eg. if a body is frozen, it might not need transform) and do a call to get these bodies transforms in bulk. Then, update the bodies state (modify the `_sync_body_state` function to instead of getting pointer to body_state to get the actual data it needs).

For this to work, there needs to be a place where we would store a reference to these nodes as to be able to iterate them quickly (eg. right now the flush_queries is called from main.cpp file. We would need to then make another call to some other service to send these info there(eg. get the nodes and update them)). Thile this might be a bit comlicated to implement, the benefits would be quite big for performance gains.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

A workaround for a person extending the physics server would be to offer such a function to get active bodies and a function to get the body state(eg. transform, velocity, etc). and then have them update the nodes manually. This is tricky as right now there is no way to disable node updates from outside afaik.

Eg. in 2d to do:
```cpp
set_block_transform_notify(true);
set_global_transform(p_state->get_transform());
set_block_transform_notify(false);
```

Basically update the rigidbody or other such nodes properties without also propagating a physics update event. Another workaround woudl be to disable the physics server and put it in some mode for these updates, but all of this would be complicated for a workaround.

### Is there a reason why this should be core and not an add-on in the asset library?

Very hard to do in an addon, needs to be core.

**Comment by Ughuuu** [comment]
As a means of providing data for this, I am benchmarking using Instruments app on mac to get the 2 largest times, `step` and `flush_queries`. I will focus on 3D for this.

## Godot: goes up to about 3000 objects

<img width="700" alt="image" src="https://github.com/user-attachments/assets/3f4bed5f-2414-43a9-9c37-312c5f1acfd8">

  - `step function` 21%
  - `flush_queries` 6%


## Jolt: goes up to about 7000 objects

<img width="567" alt="image" src="https://github.com/user-attachments/assets/8eb126e6-2dab-4a0f-ac71-f0c96736abb7">

10% step and 10% flush (am running without dev as there is some issue with latest jolt and latest godot).

## Rapier: goes up to about 5000 objects

<img width="446" alt="image" src="https://github.com/user-attachments/assets/782f4009-c21e-4bfa-9fad-5a7001cda41c">

17% step and 37% flush. Note this is written with godot-rust so the bottleneck of the callable and everything is even higher.

**Comment by reduz** [comment]
The problem is not the callback, the problem is that the whole scene system is not optimal to work like this. If you have so many objects use the servers directly and bypass the scene.

**Comment by reduz** [comment]
I still think it could be a good idea to profile what is going on at the scene side, though, there may be some silly bottlenecks there, but I think whathever it is, a single C function pointer callback should never have this impact.

**Comment by mihe** [comment]
It's also worth keeping in mind that if eliminating the `Callable` overhead is the goal here, then you also need to consider the overhead of interacting with Godot collections like `Packed*Array` from GDExtension, since pretty much every interaction with them (including `append` and `operator[]`) will have to cross the extension boundary (i.e. invoke a function pointer). I'm not sure how godot-rust implements `Callable` though, so it might still make a sizeable difference there, but for godot-cpp I wouldn't expect to see as big of a difference between the two approaches, assuming you're still fetching the same state.

> iterate through all RID's and match then with actual object
> [...]
> For this to work, there needs to be a place where we would store a reference to these nodes

As talked about on Rocket Chat, the way you'd do this right now is to ask the physics server for the body's instance ID (`ObjectID`) and then fetch the corresponding node with that using `ObjectDB::get_instance`, which will involve some non-trivial overhead as well.

On top of that you need to somehow filter out the nodes that still need to have their `_body_state_changed` method invoked, since you still have things like `_integrate_forces` to invoke and contact signals to emit, for every active `RigidBody*D`. On the 3D side you also have `PhysicalBone3D` which similarly needs to be filtered out and have its specific `_body_state_changed` method invoked. The naive approach to this would be to use `Object::cast_to` I guess, but that involves `dynamic_cast`, which is typically quite slow.

**Comment by Ughuuu** [comment]
> I still think it could be a good idea to profile what is going on at the scene side, though, there may be some silly bottlenecks there, but I think whathever it is, a single C function pointer callback should never have this impact.

I see, interesting approach. Will check this also then and see where the bottleneck comes from.

**Comment by Ughuuu** [comment]
> It's also worth keeping in mind that if eliminating the `Callable` overhead is the goal here, then you also need to consider the overhead of interacting with Godot collections like `Packed*Array` from GDExtension, since pretty much every interaction with them (including `append` and `operator[]`) will have to cross the extension boundary (i.e. invoke a function pointer). I'm not sure how godot-rust implements `Callable` though, so it might still make a sizeable difference there, but for godot-cpp I wouldn't expect to see as big of a difference between the two approaches, assuming you're still fetching the same state.


That's a good point, maybe the overhead in rust comes from the way the gdextension works:
flush_queries -> godot_rust(callables) -> godot -> godot_rust (get_data)

For these things, godot_rust needs to bind mutable self, and there are some things that add up. While if there was just:
flush_queries -> godot_rust (get_data) -> do stuff

Would be better in that regard.

**Comment by Ughuuu** [comment]
> As talked about on Rocket Chat, the way you'd do this right now is to ask the physics server for the body's instance ID (`ObjectID`) and then fetch the corresponding node with that using `ObjectDB::get_instance`, which will involve some non-trivial overhead as well.
> 
> On top of that you need to somehow filter out the nodes that still need to have their `_body_state_changed` method invoked, since you still have things like `_integrate_forces` to invoke and contact signals to emit, for every active `RigidBody*D`. On the 3D side you also have `PhysicalBone3D` which similarly needs to be filtered out and have its specific `_body_state_changed` method invoked. The naive approach to this would be to use `Object::cast_to` I guess, but that involves `dynamic_cast`, which (last I checked) is quite slow.


I think I would still leave that integrate_forces function, I would just change active_body callback.

Not 100% sure about it, but at least on the callstack (for godot-rapier) I do see some `dynamic_casts` happening in the investigation anyway.

**Comment by mihe** [comment]
> For these things, godot_rust needs to bind mutable self, and there are some things that add up. While if there was just:
> flush_queries -> godot_rust (get_data) -> do stuff

Less work is better, of course, but given the `Object`-agnostic design of the server architecture in Godot I feel like there's always going to be one or two relatively expensive steps added before "do stuff" there.

> I think I would still leave that integrate_forces function, I would just change active_body callback.

Right, but `_integrate_forces` (and the contact signals) are invoked as part of that callback, as seen [here](https://github.com/godotengine/godot/blob/03afb92efa18874da19f7fc185a32c005d20aa1d/scene/2d/physics/rigid_body_2d.cpp#L170). So I would imagine your options are then to either keep invoking the callback from the physics server (somewhat defeating the point of all of this) or to have Godot invoke `_body_state_changed` (and thus `_integrate_forces` and the signals) directly on the bodies it gets from this new `space_get_active_bodies`.

That last option would presumably involve some kind of filtering/downcasting, since you'll have bodies of type `BODY_MODE_RIGID`, `BODY_MODE_RIGID_LINEAR` and `BODY_MODE_KINEMATIC` in that list. Not only that, some of the bodies could have been added directly through the physics server by the end-user, and their respective instance IDs might reference script-defined classes, or might not even be `ObjectID`s at all, and instead just be some index into an array somewhere.

> I do see some `dynamic_casts` happening in the investigation anyway

`Object::cast_to` is very prevalent in Godot.

**Comment by Pizzaandy** [comment]
I'm developing a [Box2D v3 plugin ](https://github.com/Pizzaandy/godot-box2d-v3) and came across this issue during profiling. Box2D is so fast that state sync callbacks usually take more time than the simulation. 

Using Tracy, I found that a fair amount of time is spent just updating transforms. I'm not sure I see anything that I'd call a "silly bottleneck." 

Note: State sync callbacks are only 1.49% of frame time because I'm using Godot Physics here. Testing with 5000 RigidBody2D nodes.

Self time:

![Image](https://github.com/user-attachments/assets/18729f68-d838-4e84-a81f-8da5a1a1e660)

Total time:

![Image](https://github.com/user-attachments/assets/2c7032f7-b1f9-4eeb-a0b5-08ef6d9238df)

**Comment by Ughuuu** [comment]
Honestly, in the rapier plugin I gave up on this and just ended up disabling the state update for every object https://godot.rapier.rs/docs/documentation/performance :
1. I exposed a `RapierPhysicsServer.body_set_state_sync_callback` so that I could do:

```cpp
RapierPhysicsServer.body_set_state_sync_callback(body_rid, Callable())
```
On every object (not great, but if want performance with this I gained about 2x performance boost).
2. I added 2 more functions for getting stuff needed for rendering:

```cpp
var body_rids: Array = RapierPhysicsServer2D.space_get_active_bodies(space_rid)
var body_transforms: Array[Transform2D] = RapierPhysicsServer2D.space_get_bodies_positions(body_rids)
#or
var body_rids: Array = RapierPhysicsServer3D.space_get_active_bodies(space_rid)
var body_transforms: Array[Transform3D] = RapierPhysicsServer3D.space_get_bodies_positions(body_rids)
```

And then in GDScript went ahead and rendered them raw:

```cpp
extends MultiMeshInstance2D

func _process(delta: float) -> void:
	var bodies := RapierPhysicsServer2D.space_get_active_bodies(get_viewport().world_2d.space)
	var transforms := RapierPhysicsServer2D.space_get_bodies_transform(get_viewport().world_2d.space, bodies)
	multimesh.instance_count = transforms.size()
	var idx = 0
	for transform in transforms:
		multimesh.set_instance_transform_2d(idx, transform)
		idx += 1
```

Doing this I got on my specs about 2x speed increase, so obviously a bottleneck was eliminated, but hard to say which.

**Comment by Ughuuu** [comment]
> I'm developing a [Box2D v3 plugin ](https://github.com/Pizzaandy/godot-box2d-v3) and came across this issue during profiling. Box2D is so fast that state sync callbacks usually take more time than the simulation.
> 
> Using Tracy, I found that a fair amount of time is spent just updating transforms. I'm not sure I see anything that I'd call a "silly bottleneck."
> 
> Note: State sync callbacks are only 1.49% of frame time because I'm using Godot Physics here. Testing with 5000 RigidBody2D nodes.
> 
> Self time:
> 
> ![Image](https://github.com/user-attachments/assets/18729f68-d838-4e84-a81f-8da5a1a1e660)
> 
> Total time:
> 
> ![Image](https://github.com/user-attachments/assets/2c7032f7-b1f9-4eeb-a0b5-08ef6d9238df)

A bit confused about this, I thought the bottleneck was the flush function, but in here it appears as if the bottleneck is the get_global_transforms? Is that the root of the call?

**Comment by Pizzaandy** [comment]
Sorry, those stats were only for the code that runs inside the body state callback. I've added a few extra zones (including `GodotPhysicsServer2D::flush_queries`) for reference here. 

![Image](https://github.com/user-attachments/assets/f43bc7a3-55a6-4cbc-9895-9a5e8da43a11)

While overhead from Callables does take up most of the time (specifically checking `is_valid`), `set_global_transform` also has a relatively high amount of self time for some reason.

Unfortunately I'm not using the latest Tracy versions with a flame graph - but here's a representative breakdown of a single call to `GodotBody2D::call_queries`:

![Image](https://github.com/user-attachments/assets/b81e8808-3c91-4abd-8142-d6bcd48ee139)

**Comment by Ughuuu** [comment]
@reduz can you please take another look? I'm not sure about the use Physics Server directly workaround. If that were the case, then the nodes would not be used at all. The whole flow could be simplified a lot if some of these issues would be solved, and the benefit would be huge

**Note**:
Unity is integrating box2d v3 in their engine, and theirs will be much faster than Godot,  since Godot has this issue. Even if the advanced use case doesn't have the bottleneck, in the eyes of regular people, Godot will be slower, which right now it is with the bottlenecks said above. They achieve with box2d 25K joints and 25K bodies: https://www.youtube.com/watch?v=TXlLRagoYyQ . In Godot with regular nodes this is impossible, only with MeshInstance for rendering, and disabling all the basic callbacks of nodes (if you want to use nodes, but then you will lose all the nice scripting features, etc.)

**Note2**:
Unity is also integrating the determinism aspect of box2d into their engine: https://www.youtube.com/watch?v=PID-TcuiD9w

I just think that your statement in this issue makes it seem like if there is no issue, but I believe that to not be the case, and that big improvements can be made.

**Comment by KeyboardDanni** [comment]
> The problem is not the callback, the problem is that the whole scene system is not optimal to work like this. If you have so many objects use the servers directly and bypass the scene.

The server direct query functions could use some optimization as well: https://github.com/godotengine/godot-proposals/issues/4474

It's also a question of ease of use. Direct server access really should be a last resort when easier user-facing options have already been exhausted.
